### ISSUE TITLE
feature proposal: iterable interfaces

### ISSUE THREAD

**Issue by derhuerst** [w8r/avl#23]
Using [iterables](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) to walk over a large amount of items/data has two advantages:

- In contrast to computing a full array all items, items can be read from the tree on the fly. This prevents unnecessary work.
- Iterables are well-supported within ECMAScript and libs: `String`, `Array`, `Map`, `Set` etc. are iterable.

Examples of how iterables would be useful for this lib:

```js
for (const key of tree.keys()) {
  await userRequestedAKey()
  console.log(key)
}

for (const node of tree.range('A', 'O')) {
  console.log(node.data)
}
```

**Comment by derhuerst** [comment]
I'd be happy to submit a PR with this!

**Comment by w8r** [comment]
Please do! Would that still be es5 compatible?

**Comment by derhuerst** [comment]
> Would that still be es5 compatible?

Nope, [`Symbol.iterator` is ES2015/ES6](http://kangax.github.io/compat-table/es6/#test-well-known_symbols_Symbol.iterator,_existence_a_href=_https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator_title=_MDN_documentation_img_src=_../mdn.png_alt=_MDN_(Mozilla_Development_Network)_logo_width=_15_height=_13_/_/a_nbsp;), so not supported in IE11.

But I think You can implement it in a backwards-compatible way:

```js
if (Symbol && Symbol.iterator) {
  AVLTree.prototype[Symbol.iterator] = function (...) {...}
}
```

**Comment by w8r** [comment]
That would be cool!

**Comment by derhuerst** [comment]
What API do you prefer?

1. `tree` is not iterable
2. `tree[Symbol.iterator]` returns a *keys* iterable?
3. `tree[Symbol.iterator]` returns an *entries* (`[key, value]`) iterable?

1. `tree.keys()` stays an array as before, add `tree.keysIterable()`.
2. Change `tree.keys()` from an array to an iterable. (breaking)

**Comment by w8r** [comment]
3. tree[Symbol.iterator] returns an entries ([key, value]) iterable?

--

1. `tree.keys()` stays an array, as before, add `tree.keysIterable()`

right?

**Comment by derhuerst** [comment]
Okay, will implement this. A `tree.valuesIterable()` also makes sense, right?

**Comment by w8r** [comment]
yes, let's do it like that. Maybe I will change the API later. I will also add your changes to the `splay-tree` repo then

**Comment by derhuerst** [comment]
I have implemented this at [`derhuerst/avl#iterables`](https://github.com/derhuerst/avl/tree/iterables), but using ES6 generators. [They are not supported in IE11](https://caniuse.com/#feat=es6-generators), so we have two choices now:

- implement the iterators by hand using only ES5 syntax,
- or use Babel for transpiling or get bublé to transpile generators.

What do you prefer?
